### PR TITLE
import from the new location of `normalize_axis_index` if possible

### PR DESCRIPTION
**Pull request by keewis**
Another one of the `numpy=2.0` fixes, this time `numpy.core.multiarray.normalize_axis_index` has been moved to `numpy.lib.array_utils` (and apparently this is the first time it has been officially exposed as public API).

Since as far as I remember `numpy` is working on removing `numpy.core` entirely, we might also want to change our usage of `defchararray` (in the formatting tests). Not sure how, though.

- [x] Towards #8091

### CODE REVIEW

**Comment by keewis** [comment]
it appears we have an issue with `cftime` being built against `numpy<2.0` but used with `numpy>=2.0`. Not sure if we can do anything about that, as `cftime` build-depends on `oldest-supported-numpy`, which for `python=3.11` pins `numpy=1.23.2`.

`numpy` helpfully prints that we should either use wheels built against `numpy>=2.0` or disable build isolation when having `pip` build the wheels from source. In the most recent commit I did the latter, let's see if that worked.

**Comment by keewis** [comment]
it appears the reason for the failing `iris` tests is `cf_units` and the ones for the failing backends tests are `netcdf4` and `h5py` which would all need to be built with `numpy>=2.0.0.dev0`.

Does anyone have any ideas? Otherwise we'd have to live with our upstream-dev CI failing until compatible wheels appear on `conda-forge`.

**Comment by dcherian** [comment]
> he ones for the failing backends tests are netcdf4 and h5py which would all need to be built with numpy>=2.0.0.dev0.

I doubt these will move before a numpy release candidate.

**Comment by keewis** [comment]
I guess that means we have two choices:
1. ignore / skip the failing tests until these packages are updated
2. build ourselves in CI

where 2 would mean that we'd have to deal with compiling all three packages (plus make sure the `cftime` build from this PR works correctly)

**Comment by mathause** [comment]
For 2. the numpy recommendations seems to be to pass `--no-build-isolation` to `pip` but then you have to manually add all build dependencies and it does not correctly set the version number, which causes other problems. And there seems to be no way to force build dependencies (or their version) directly from pip. So I gave up on trying to get the upstream CI to work (for regionmask, but I guess this also applies here, unless I misunderstand something).

**Comment by keewis** [comment]
we don't install dev `netcdf4` or `hdf5` / `h5py` (as far as I understand it, `h5netcdf` is pure-python), yet, so unless we change that it would become just "remove `netcdf4` and `hdf5` until they support `numpy` dev", which is much easier

**Comment by dcherian** [comment]
Works for me, since we know numpy is making major changes.

**Comment by keewis** [comment]
it appears the recent commits fix most of the issues we had with the incompatibility, and we're down to ~15 failing tests now. Thus, I'll mark this as ready and I'll have this PR close the current nightly failures issue, since that has been lingering for way too long.

**Comment by keewis** [comment]
I'll note that `flox` appears to be using several functions from `numpy.core`, as well, so we'll have to update that, too.

File: xarray/core/duck_array_ops.py [hunk 1]
```diff
@@ -39,6 +38,13 @@
 from xarray.core.pycompat import array_type, is_duck_dask_array
 from xarray.core.utils import is_duck_array, module_available
 
+# remove once numpy 2.0 is the oldest supported version
+try:
```
[the remark this answers is not in the document]
Comment (keewis) [reply xarray/core/duck_array_ops.py:42]: does that capture the current dev versions of `numpy` that are technically before `numpy=2.0.0`? If yes that sounds good to me.

[the remark this answers is not in the document]
Comment (dcherian) [reply xarray/core/duck_array_ops.py:42]: maybe not, I guess we could specify `1.90.0`?

[the remark this answers is not in the document]
Comment (keewis) [reply xarray/core/duck_array_ops.py:42]: to answer my own question, we can use `2.0.0.dev0` as the min-version:
```python
from packaging import version

version.parse("2.0.0") >= version.parse("2.0.0.dev0")  # True
version.parse("2.0.0rc1") >= version.parse("2.0.0.dev0")  # True
```

**Comment by keewis** [comment]
`mypy` needs some work (probably ignores?), but otherwise this should be ready

**Comment by Illviljan** [comment]
```
xarray/core/nputils.py:15: error: Cannot find implementation or library stub for module named "numpy.lib.array_utils"  [import-not-found]
xarray/core/duck_array_ops.py:43: error: Cannot find implementation or library stub for module named "numpy.lib.array_utils"  [import-not-found]
xarray/core/duck_array_ops.py:43: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```
Numpy has these odd errors sometimes, I haven't fully figured them out yet. Ignore if rephrasing the code doesn't work either.

**Comment by dcherian** [comment]
Thanks @keewis !

**Pull request merged**